File: 3-Strings/ex1253.c
#include <stdio.h>
#include <string.h>

void decodificador(char frase[], int desloc, char fraseDecodificada[]){
    char caracter;
    char alfabeto[27] = {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'};
    int i, tam = strlen(frase), indice, j, pos;

    for(i = 0; i < tam; i++){
        caracter = frase[i];
        for(j = 0; j < 27; j++){
            if(caracter == alfabeto[j]){
                indice = j;
                break;
            }
        }
        pos = indice - desloc;
        if(pos < 0){
            pos = 27 - (desloc - indice + 1);
        } 
        fraseDecodificada[i] = alfabeto[pos];
    }
    fraseDecodificada[i] = '\0';
}


int main(){
    //variaveis
    int n, i, desloc = 0;
    char frase[51], fraseDecodificada[51];

    //entrada
    scanf("%d", &n);

    for(i = 0; i < n; i++){
        fflush(stdin);
        scanf("%s", frase);
        scanf("%d", &desloc);

        //processamento
        decodificador(frase, desloc, fraseDecodificada);

        //saida
        printf("%s\n", fraseDecodificada);
    }
    return 0;
}

File: 3-Strings/ex1253.py
def decodificador(frase, desloc):
    tam = len(frase)
    fraseDecodificada = str()
    for i in range(0, tam):
        codAscii = ord(frase[i])
        caracter = codAscii - desloc
        if caracter < 65:
            caracter = 90 - (65 - (codAscii - desloc) - 1)
        fraseDecodificada += chr(caracter)
    print(fraseDecodificada)

#entrada
n = int(input())

for i in range(0, n):
    frase = str(input())
    desloc = int(input())

    #processamento & saida
    decodificador(frase, desloc)
